File: Lecture-8-Build-Netflix-Recommendation-System/code/pythonDemo/src/pythondemo/embedding.py
from typing import Protocol

from openai import OpenAI


class EmbeddingClient(Protocol):
    def embed(self, text: str) -> list[float]: ...


class OpenAIEmbeddingClient:
    def __init__(self, client: OpenAI, model: str):
        self._client = client
        self._model = model

    def embed(self, text: str) -> list[float]:
        response = self._client.embeddings.create(model=self._model, input=text)
        return response.data[0].embedding

File: Lecture-8-Build-Netflix-Recommendation-System/code/goDemo/internal/embedding/embedding.go
package embedding

import (
	"context"

	"github.com/openai/openai-go/v2"
)

// Client abstracts away the embedding provider, so MovieService does not
// depend on OpenAI directly.
type Client interface {
	Embed(ctx context.Context, text string) ([]float64, error)
}

// OpenAIClient is a Client backed by the OpenAI embeddings API.
type OpenAIClient struct {
	client openai.Client
	model  string
}

func NewOpenAIClient(client openai.Client, model string) *OpenAIClient {
	return &OpenAIClient{client: client, model: model}
}

func (c *OpenAIClient) Embed(ctx context.Context, text string) ([]float64, error) {
	response, err := c.client.Embeddings.New(ctx, openai.EmbeddingNewParams{
		Input: openai.EmbeddingNewParamsInputUnion{OfString: openai.String(text)},
		Model: openai.EmbeddingModel(c.model),
	})
	if err != nil {
		return nil, err
	}

	return response.Data[0].Embedding, nil
}
